contracts/libraries/SwapToRatio.sol

```diff
@@ -0,0 +1,189 @@
+        } else {
+            amount0Updated = positionParams.amount0Initial + token0Delta;
+            amount1Updated = positionParams.amount1Initial - ((token1Delta * 1e6) / (1e6 - poolParams.fee));
+            doSwap = (amount1Updated) / amount0Updated > (validDeposit1) / validDeposit0;
+        }
+    }
+
+    // TODO: address range order scenarios
+    function getPostSwapPrice(IUniswapV3Pool pool, PositionParams memory positionParams)
+        internal
+        view
+        returns (uint160)
+    {
+        (PoolParams memory poolParams, int24 tickSpacing, int24 tick) = getPoolInputs(pool);
+        bool zeroForOne =
+            isZeroForOne(
+                positionParams.amount0Initial,
+                positionParams.amount1Initial,
+                poolParams.sqrtRatioX96,
+                positionParams.sqrtRatioX96Lower,
+                positionParams.sqrtRatioX96Upper
+            );
+        bool crossTickBoundary = true;
+        uint256 amount0Next;
+        uint256 amount1Next;
+        int24 nextInitializedTick;
+        uint160 sqrtRatioNextX96;
+
+        while (crossTickBoundary) {
+            // returns the next initialized tick or the last tick within one word of the current tick.
+            // We'll renew calculation at least on a per word basis for better rounding
+            (nextInitializedTick, ) = pool.nextInitializedTickWithinOneWord(tick, tickSpacing, zeroForOne);
+            if (zeroForOne) nextInitializedTick -= 1;
+            sqrtRatioNextX96 = TickMath.getSqrtRatioAtTick(nextInitializedTick);
+
+            (crossTickBoundary, amount0Next, amount1Next) = swapToNextInitializedTick(
+                poolParams,
+                positionParams,
+                sqrtRatioNextX96,
+                zeroForOne
+            );
+
+            // if crossing an initialized tick, update token amounts and other parameters to values at next tick
+            if (crossTickBoundary) {
+                console.log('nextTick');
+                console.logInt(nextInitializedTick);
+                (, int128 liquidityNet, , , , , , ) = pool.ticks(nextInitializedTick);
+
+                positionParams.amount0Initial = amount0Next;
+                positionParams.amount1Initial = amount1Next;
+                poolParams.sqrtRatioX96 = sqrtRatioNextX96;
+                // overflow desired
+                poolParams.liquidity += uint128(liquidityNet);
+                tick = nextInitializedTick;
+            }
+        }
+        return sqrtRatioNextX96;
+        // TODO: return calculateConstantLiquidityPostSwapSqrtPrice
+        // return calculateConstantLiquidityPostSwapSqrtPrice(poolParams, positionParams);
+    }
+
+    function isZeroForOne(
+        uint256 amount0,
+        uint256 amount1,
+        uint160 sqrtRatioX96,
+        uint160 sqrtRatioX96Lower,
+        uint160 sqrtRatioX96Upper
+    ) internal view returns (bool) {
```

Comment:
for precision purposes, we might consider doing some of:

1. transform this from a/b > c/d to a > (c*b)/d, where the RHS uses mulDiv
2. use a much larger liquidity value (as large as possible while avoiding overflow)
3. think through whether we should round c and d in different directions, to bias the result in one direction or the other (would need to think through this one)

some/all of these might require bounding amounts to uint128 or similar